homeassistant/components/weatherflow/config_flow.py

```diff
@@ -0,0 +1,88 @@
+"""Config flow for WeatherFlow."""
+from __future__ import annotations
+
+import asyncio
+from asyncio import Future
+from typing import Any
+
+# import my_pypi_dependency
+from pyweatherflowudp.client import EVENT_DEVICE_DISCOVERED, WeatherFlowListener
+from pyweatherflowudp.const import DEFAULT_HOST
+from pyweatherflowudp.errors import AddressInUseError, ListenerError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import DOMAIN, LOGGER
+
+STEP_USER_DATA_SCHEMA = vol.Schema({vol.Required(CONF_HOST, default=DEFAULT_HOST): str})
+
+
+async def _async_has_devices(hass: HomeAssistant) -> bool:
+    """Return if there are devices that can be discovered."""
+
+    future_event: Future[bool] = asyncio.Future()
+    host: str = DEFAULT_HOST
+
+    @callback
+    def found():
+        """Handle a discovered device."""
+        future_event.set_result(True)
+
+    async with WeatherFlowListener(host) as client:
+        LOGGER.debug("Registering EVENT_DISCOVERED_FUNCTION")
+        try:
+            client.on(EVENT_DEVICE_DISCOVERED, lambda _: found())
+            await asyncio.wait_for(future_event, timeout=10)
```

Comment:
use asyncio.timeout instead

Reply:
I can't seem to make it work with asyncio.timeout.  I tried the following code block:

```python
        try:
            with asyncio.timeout(10):
                client.on(EVENT_DEVICE_DISCOVERED, lambda _: _async_found())
                await future_event.result()
        except asyncio.TimeoutError:
            return False
        return True
```

>     with asyncio.timeout(10):
> TypeError: 'Timeout' object does not support the context manager protocol

Reply:
`async with asyncio.timeout`

Reply:
Thanks...

This seems to be working but I removed the `_async_found()` and kept it as `found()` because I dont think any async stuff is happening.



```
async def _async_can_discover_devices(host: str = DEFAULT_HOST) -> bool:
    """Return if there are devices that can be discovered."""

    future_event: Future[bool] = asyncio.get_running_loop().create_future()

    @callback
    def _found():
        """Handle a discovered device."""
        future_event.set_result(True)
        return future_event

    async with WeatherFlowListener(host) as client:
        LOGGER.debug("Registering EVENT_DISCOVERED_FUNCTION")
        async with asyncio.timeout(10):
            try:
                client.on(EVENT_DEVICE_DISCOVERED,  lambda _: _found())
                await future_event
            except asyncio.TimeoutError:
                return False
    return True
```

Reply:
... or I liked perhaps and its not working actually :(

Reply:
I'm just struggling without this call:

`await asyncio.wait_for(future_event, timeout=10)` to get it to work correctly.. I'll keep trying

Reply:
... I think I tracked down the error...  as multiple devices get found we try to set the future result once again and its already set so I get an invalid state...

Reply:
Looks like because multiple devices get found -> it was trying to call set_result on a finished future ... so that was causing an issue... (will push soon)